File: Code/Source/Estimatorx.Web/app/services/ProjectCalculator.ts
/// <reference path="../_ref.ts" />

module Estimatorx {
    "use strict";

    export class ProjectCalculator {
        static $inject = [];

        constructor() {

        }

        updateTotals(project: IProject) {
            var hoursPerWeek = project.HoursPerWeek || 30;

            // reset
            project.TotalTasks = 0;
            project.TotalHours = 0;
            project.TotalWeeks = 0;
            project.ContingencyHours = 0;
            project.ContingencyWeeks = 0;

            angular.forEach(project.Sections, (section) => {
                this.updateSection(project, section);

                // update parent totals
                project.TotalTasks += section.TotalTasks;
                project.TotalHours += section.TotalHours;

            });

            // calculate weeks after children total
            project.TotalWeeks = this.round(project.TotalHours / hoursPerWeek, 2);

            var contingencyPercent = project.ContingencyRate / 100;
            var contingencyHours = project.TotalHours * (1 + contingencyPercent);
            project.ContingencyHours = Number(this.round(contingencyHours, 0));
            project.ContingencyWeeks = this.round(project.ContingencyHours / hoursPerWeek, 2);

            return project;
        }

        updateSection(project: IProject, section: ISection) {
            // reset
            section.TotalTasks = 0;
            section.TotalHours = 0;
            section.TotalWeeks = 0;

            angular.forEach(section.Tasks, (estimate) => {
                this.updateEstimate(project, estimate);

                // update parent totals
                section.TotalTasks += estimate.TotalTasks;
                section.TotalHours += estimate.TotalHours;
            });

            // calculate weeks after children total
            var hoursPerWeek = project.HoursPerWeek;
            section.TotalWeeks = this.round(section.TotalHours / hoursPerWeek, 2);
        }

        updateEstimate(project: IProject, task: ITask) {
            var verySimple = task.VerySimple || 0,
                simple = task.Simple || 0,
                medium = task.Medium || 0,
                complex = task.Complex || 0,
                veryComplex = task.VeryComplex || 0;


            task.TotalTasks = verySimple + simple + medium + complex + veryComplex;

            var factor = <IFactor>Enumerable.From(project.Factors)
                .Where(f => f.Id == task.FactorId)
                .FirstOrDefault(<IFactor>{});

            var factorVerySimple = factor.VerySimple || 0,
                factorSimple = factor.Simple || 0,
                factorMedium = factor.Medium || 0,
                factorComplex = factor.Complex || 0,
                factorVeryComplex = factor.VeryComplex || 0;

            task.TotalHours = (verySimple * factorVerySimple)
                + (simple * factorSimple)
                + (medium * factorMedium)
                + (complex * factorComplex) 
                + (veryComplex * factorVeryComplex);

        }

        round(value: number, decimals: number = 2): number {
            if (decimals <= 0)
                return Math.round(value);

            var m: number = Math.pow(10, decimals);
            return Math.round(value * m) / m;
        }
    }


    // register service
    angular.module(Estimatorx.applicationName)
        .service('projectCalculator', ProjectCalculator);

}

File: migration/Estimatorx.Data.Mongo/ProjectCalculator.cs
using System;
using System.Linq;

namespace Estimatorx.Data.Mongo
{
    public static class ProjectCalculator
    {
        /// <summary>
        /// Updates all the totals the specified <paramref name="project"/>.
        /// </summary>
        /// <param name="project">The project to total.</param>
        /// <returns></returns>
        public static Project UpdateTotals(Project project)
        {
            var hoursPerWeek = (double)project.HoursPerWeek;

            foreach (var section in project.Sections)
            {
                UpdateSection(project, section);

                // update parent totals
                project.TotalTasks += section.TotalTasks;
                project.TotalHours += section.TotalHours;
            }

            // calculate weeks after children total
            project.TotalWeeks = Math.Round(project.TotalHours / hoursPerWeek, 2);

            double contingencyPercent = project.ContingencyRate / 100;
            double contingencyHours = project.TotalHours * (1 + contingencyPercent);
            project.ContingencyHours = Convert.ToInt32(Math.Round(contingencyHours, 0));
            project.ContingencyWeeks = Math.Round(project.ContingencyHours / hoursPerWeek, 2);

            return project;
        }

        private static void UpdateSection(Project project, Section section)
        {
            foreach (var task in section.Tasks)
            {
                UpdateEstimate(project, task);

                // update parent totals
                section.TotalTasks += task.TotalTasks;
                section.TotalHours += task.TotalHours;
            }

            // calculate weeks after children total
            var hoursPerWeek = (double)project.HoursPerWeek;
            section.TotalWeeks = Math.Round(section.TotalHours / hoursPerWeek, 2);
        }

        private static void UpdateEstimate(Project project, Task task)
        {
            byte verySimple = task.VerySimple ?? 0;
            byte simple = task.Simple ?? 0;
            byte medium = task.Medium ?? 0;
            byte complex = task.Complex ?? 0;
            byte veryComplex = task.VeryComplex ?? 0;

            var factor = project.Factors.FirstOrDefault(f => f.Id == task.FactorId) ?? new Factor();

            task.TotalTasks = verySimple
                + simple
                + medium
                + complex
                + veryComplex;

            task.TotalHours = verySimple * factor.VerySimple
                + simple * factor.Simple
                + medium * factor.Medium
                + complex * factor.Complex
                + veryComplex * factor.VeryComplex;
        }
    }
}
